spot_ros2_control/hardware/src/spot_hardware_interface.cpp

```diff
@@ -0,0 +1,367 @@
+  if (joint_pos.size() == 0 || joint_vel.size() == 0 || joint_load.size() == 0) {
+    return hardware_interface::return_type::OK;
+  }
+  // Ensure that the states received from the Spot SDK will fit into the hw_states_ vector
+  const auto states_size = hw_states_.size();
+  if (interfaces_per_joint_ * joint_pos.size() != states_size ||
+      interfaces_per_joint_ * joint_vel.size() != states_size ||
+      interfaces_per_joint_ * joint_load.size() != states_size) {
+    RCLCPP_FATAL(
+        rclcpp::get_logger("SpotHardware"),
+        "The number of joints and interfaces does not match with the outputted joint states from the Spot SDK!");
+    return hardware_interface::return_type::ERROR;
+  }
+  // Read values into joint states
+  for (uint i = 0; i < states_size; i += interfaces_per_joint_) {
+    hw_states_.at(i) = joint_pos.at(i / interfaces_per_joint_);
+    hw_states_.at(i + 1) = joint_vel.at(i / interfaces_per_joint_);
+    hw_states_.at(i + 2) = joint_load.at(i / interfaces_per_joint_);
+  }
+  return hardware_interface::return_type::OK;
+}
+
+hardware_interface::return_type SpotHardware::write(const rclcpp::Time& /*time*/, const rclcpp::Duration& /*period*/) {
+  // This function will be responsible for sending commands to the robot via the BD SDK -- currently unimplemented.
+  return hardware_interface::return_type::OK;
+}
+
+bool SpotHardware::authenticate_robot(const std::string& hostname, const std::string& username,
+                                      const std::string& password) {
+  // Create a Client SDK object.
+  const auto client_sdk = ::bosdyn::client::CreateStandardSDK("SpotHardware");
+  auto robot_result = client_sdk->CreateRobot(hostname);
+  if (!robot_result) {
+    RCLCPP_INFO(rclcpp::get_logger("SpotHardware"), "Could not create robot");
+    return false;
+  }
+  robot_ = robot_result.move();
+  const ::bosdyn::common::Status status = robot_->Authenticate(username, password);
+  if (!status) {
+    RCLCPP_INFO(rclcpp::get_logger("SpotHardware"),
+                "Could not authenticate robot with hostname: %s username: %s password: %s", hostname.c_str(),
+                username.c_str(), password.c_str());
+    return false;
+  }
+  RCLCPP_INFO(rclcpp::get_logger("SpotHardware"), "Robot successfully authenticated!");
+  return true;
+}
+
+bool SpotHardware::start_time_sync() {
+  // Establish time synchronization with the robot
+  ::bosdyn::client::Result<::bosdyn::client::TimeSyncClient*> time_sync_client_resp =
+      robot_->EnsureServiceClient<::bosdyn::client::TimeSyncClient>();
+  if (!time_sync_client_resp.status) {
+    RCLCPP_INFO(rclcpp::get_logger("SpotHardware"), "Could not create time sync client");
+    return false;
+  }
+  RCLCPP_INFO(rclcpp::get_logger("SpotHardware"), "Created time sync client");
+  ::bosdyn::client::TimeSyncClient* time_sync_client = time_sync_client_resp.response;
+  ::bosdyn::client::TimeSyncThread time_sync_thread(time_sync_client);
```

Comment:
@khughes-bdai shouldn't `time_sync_thread` outlive its scope here? As it stands, the thread will be stopped when `start_time_sync` returns.

Reply:
Hmm, I took a lot of this logic from the spot-rl repo which doesn't do this either:
https://github.com/bdaiinstitute/spot-rl/blob/cfee0ae675b2abbbc8d233322de391afe2036543/cpp/src/spot.cpp#L55
@mbrauckmann-bdai was this ever an issue with your on-robot testing?

Reply:
Note that code gets the time sync thread from the robot, it doesn't instantiate one of its own.

Reply:
Ah that is my bad -- will update this

Reply:
Added in latest commit, the time sync thread now comes from `robot_`